### ISSUE TITLE
[pubsub] Batch-ack messages

### ISSUE THREAD

**Issue by ditsuke** [yoshidan/google-cloud-rust#49]
I'm working on a high-perf requirement pubsub client where I've been told it would be unacceptable to individually ack messages. I wasn't able to see examples for batch acking here, and from what I see going on internally in `ReceivedMessage.ack()` ([reference](https://github.com/yoshidan/google-cloud-rust/blob/main/pubsub/src/subscriber.rs#L35)), it calls the crate-private `SubscriberClient.acknowledge()` method which supports batches. Is this API exposed publicly somewhere in the package or ack-batching taken care of in some other way?

**Comment by yoshidan** [comment]
Batch execution of Ack is not currently supported.
I am considering batch acking with the following code by exposing `ReceivedMessage.ack_id` and adding an `ack` method to `Subscription`

```
let (sender, mut receiver)  = tokio::sync::mpsc::unbounded_channel();
let subscription_for_receive = subscription.clone();

// spawn receiver
let handle = tokio::spawn(async move {
    let _ = subscription_for_receive .receive(move |message, _ctx| {
        let sender = sender.clone();
        async move {
            sender.send(message.ack_id).unwrap();
        }
    }, None,None ).await;
});

// spawn ack manager
let ack_manager = tokio::spawn( async move {
    let mut ack_ids = Vec::new();
    loop {
        match tokio::time::timeout(Duration::from_secs(2), receiver.recv()).await {
            Ok(ack_id) => match ack_id {
                Some(ack_id) => {
                    ack_ids.push(ack_id);
                    if ack_ids.len() > 10 {
                        subscription.ack(ack_ids).await;
                        ack_ids = Vec::new();
                    }
                },
                None => break  // channel is closed
            },
            Err(e) => {
                // timeout
                if !ack_ids.is_empty() {
                    subscription.ack(ack_ids).await;
                    ack_ids = Vec::new();
                }
            }
        }
    }
});
```

**Comment by ditsuke** [comment]
> Batch execution of Ack is not currently supported.
> I am considering batch acking with the following code by exposing `ReceivedMessage.ack_id` and adding an `ack` method to `Subscription`
> 
> ```
> let (sender, mut receiver)  = tokio::sync::mpsc::unbounded_channel();
> let subscription_for_receive = subscription.clone();
> 
> // spawn receiver
> let handle = tokio::spawn(async move {
>     let _ = subscription_for_receive .receive(move |message, _ctx| {
>         let sender = sender.clone();
>         async move {
>             sender.send(message.ack_id);
>         }
>     }, None,None ).await;
> });
> 
> // spawn ack manager
> let ack_manager = tokio::spawn( async move {
>     let mut ack_ids = Vec::new();
>     loop {
>         match tokio::time::timeout(Duration::from_secs(2), receiver.recv()).await {
>             Ok(ack_id) => match ack_id {
>                 Some(ack_id) => {
>                     ack_ids.push(ack_id);
>                     if ack_ids.len() > 10 {
>                         subscription.ack(ack_ids).await;
>                         ack_ids = Vec::new();
>                     }
>                 },
>                 None => break  // channel is closed
>             },
>             Err(e) => {
>                 // timeout
>                 if !ack_ids.is_empty() {
>                     subscription.ack(ack_ids).await;
>                     ack_ids = Vec::new();
>                 }
>             }
>         }
>     }
> });
> ```


@yoshidan I like the approach, very natural! Please let me know if you're short on time and I'll be happy to string this up and open a PR for you to review shortly. Anything to get batch-acking possible soon.

**Comment by yoshidan** [comment]
I published the [version 0.6.0](https://crates.io/crates/google-cloud-pubsub).

Here is the example.
https://github.com/yoshidan/google-cloud-rust/blob/main/pubsub/src/subscription.rs#L365

**Comment by ditsuke** [comment]
Thank you. Trying it out!

**Comment by ditsuke** [comment]
Closed with #50
